jobs/build/ocp4/build.groovy

```diff
@@ -0,0 +1,595 @@
+        echo("Using version ${full} from latest openshift package ${prevBuild}")
+    }
+
+    if (! full.startsWith("${stream}.")) {
+        // The version we came up with somehow doesn't match what we expect to build; abort
+        error("Determined a version, '${full}', that does not begin with '${stream}.'")
+    }
+
+    stream = stream.tokenize('.').collect { it.toInteger() }
+    return [
+        major: stream[0],
+        minor: stream[1],
+        full: full,
+        release: release,
+    ]
+}
+
+/**
+ * Plan what will be built.
+ * Figure out whether we're building RPMs and/or images, and which ones, based on
+ * the parameters and which sources have changed (if relevant).
+ * Fills in the "buildPlan" property for later stages to use.
+ * @return map which is the buildPlan property of this build.
+ */
+def planBuilds() {
+    if (buildPlan.forceBuild) {
+        echo "Building all as requested regardless of whether source has changed."
+        currentBuild.description = "Building from source whether changed or not.\n"
+        return buildPlan
+    }
+
+    // otherwise we need to scan sources.
+    echo "Building only where source has changed."
+    currentBuild.description = "Building sources that have changed.\n"
+
+    def changed = [:]
+    try {
+        def yamlData = readYaml text: buildlib.doozer(
+            """
+            ${doozerOpts}
+            ${includeExclude "rpms", buildPlan.rpmsIncluded, buildPlan.rpmsExcluded}
+            ${includeExclude "images", buildPlan.imagesIncluded, buildPlan.imagesExcluded}
+            config:scan-sources --yaml
+            """, [capture: true]
+        )
+        changed = getChanges(yamlData)
+
+        def report = { msg ->
+            echo msg
+            currentBuild.description += "${msg}\n"
+        }
+        if (!buildPlan.buildRpms) {
+            report "RPMs: not building"
+        } else if (changed.rpms) {
+            report "RPMs: building " + changed.rpms.join(", ")
+            buildPlan.rpmsIncluded = changed.rpms.join(",")
+            buildPlan.rpmsExcluded = ""
+        } else {
+            report "RPMs: none changed"
+            buildPlan.buildRpms = false
+        }
+
+        if (!buildPlan.buildImages) {
+            report "Images: not building"
+            return buildPlan
+        } else if (!changed.images) {
+            report "Images: none changed"
+            buildPlan.buildImages = false
+            currentBuild.displayName += " [no changes]"
+            return buildPlan
+        }
+        report "Found ${changed.images.size()} image(s) with changes:\n  " + changed.images.join("\n  ")
+
+        // also determine child images of changed
+        yamlData = readYaml text: buildlib.doozer(
+            """
+            ${doozerOpts}
+            ${includeExclude "images", buildPlan.imagesIncluded, buildPlan.imagesExcluded}
+            images:show-tree --yml
```

Comment:
haha... I had no idea this was used for anything real. <3

Reply:
:laughing: it's now super-critical!